keanu-project/src/main/java/io/improbable/keanu/algorithms/sampling/ForwardSampler.java

```diff
@@ -0,0 +1,114 @@
+package io.improbable.keanu.algorithms.sampling;
+
+import com.google.common.collect.Sets;
+import io.improbable.keanu.KeanuRandom;
+import io.improbable.keanu.algorithms.NetworkSamples;
+import io.improbable.keanu.algorithms.Variable;
+import io.improbable.keanu.algorithms.VariableReference;
+import io.improbable.keanu.algorithms.mcmc.KeanuProposalApplicationStrategy;
+import io.improbable.keanu.algorithms.mcmc.ProposalApplicationStrategy;
+import io.improbable.keanu.algorithms.mcmc.proposal.PriorProposalDistribution;
+import io.improbable.keanu.algorithms.mcmc.proposal.Proposal;
+import io.improbable.keanu.algorithms.mcmc.proposal.ProposalDistribution;
+import io.improbable.keanu.backend.ComputableModel;
+import io.improbable.keanu.backend.KeanuComputableModel;
+import io.improbable.keanu.network.BayesianNetwork;
+import io.improbable.keanu.network.LambdaSection;
+import io.improbable.keanu.vertices.Probabilistic;
+import io.improbable.keanu.vertices.Vertex;
+import org.nd4j.base.Preconditions;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class ForwardSampler {
```

Comment:
Is there a reason this doesn't implement the PosteriorSamplingAlgorithm interface? (I know given the restrictions as they are currently the posterior == prior, but we're missing sample generation now (ie you can't stream samples) and if we relax the restrictions below the posterior could be different to the prior.

Reply:
Because it's tied to a Bayesian Network and we can't use a probabilistic model. I agree that the benefits of having it implement `PosteriorSamplingAlgorithm` are huge so I've made it implement it with the constraint you must pass Vertices. It gives us streaming for free!